### DISCUSSION TITLE
How to contribute error messages

### DISCUSSION THREAD

**Discussion by magnusviri** [pkgxdev/pkgx#487] [Help] [answered]
I think I'm going to be packaging up a lot of stuff today and the package.yml files are going to be in various degrees of chaos. I think this is an excellent opportunity to improve the error handling. But I don't know where to start. I've had a little bit of experience with TypeScript, but only for the web and this looks unlike any other codebase I've worked on. That said, I do want to learn it and figure it out.

Right now, I'd like to add a descriptive error message for this error. It's happening because the "distributable: url:" in the package.yml is 404 because the version actually needs to be version.raw.

```
> pkg build abseil.io
...
{ selected: 20230125.2.0 }
querying
undefined

error: Uncaught (in promise) Error: 404: https://github.com/abseil/abseil-cpp/archive/refs/tags/20230125.2.0.tar.gz
    throw new TeaError('http', {cause, ...opts})
          ^
    at Object.download (https://raw.githubusercontent.com/teaxyz/cli/v0.25/src/hooks/useDownload.ts:116:11)
Caused by: Error: 404: https://github.com/abseil/abseil-cpp/archive/refs/tags/20230125.2.0.tar.gz
    at internal (https://raw.githubusercontent.com/teaxyz/cli/v0.25/src/hooks/useDownload.ts:102:11)
    at async Object.download (https://raw.githubusercontent.com/teaxyz/cli/v0.25/src/hooks/useDownload.ts:108:28)
```

The error message I'm thinking could be something like, "Could not download distributable, do you need to use version.raw?". I just don't know where to put it. I know useDownload.ts line 118 is throwing the error: `throw new TeaError('http', {cause, ...opts})`. But I am pretty sure this is used by lots of code, and so that isn't the correct place for the error. I know the code calling useDownload.download is [line 88 of brewkit libexec/fetch.ts](https://github.com/teaxyz/brewkit/blob/cf18e77cd228c175ab4d539659f74f1369910733/libexec/fetch.ts#L88). But if I put that in a try/catch, it doesn't change anything, I'm guessing because this code is a promise (a design pattern I only understand 50%). Didn't work:

```
    try {
      useDownload().download({ dst, src })
    } catch {
      console.error({ dst, src })
      throw new Error(`cURL failed to download ${src}`)
    }
```

How would I catch this error and print a really helpful error message?

**Comment by mxcl** [comment, marked as answer]
Good initiative, but I just tried to answer and realized this is a shit tonne of work. Honestly your efforts would be more fruitful packaging. How about I promise to make the errors better soonish instead?

And to be fair the error message does contain all the info you need to debug currently, it's just noisy.

**Reply by magnusviri** [reply]
LOL, ok. There is no rush, I just wanted to take advantage of the plethora of errors I'm about to wade through. I will at least document all of the errors I get and their fixes, ok?

**Reply by mxcl** [reply]
k I did a quick fix https://github.com/teaxyz/brewkit/commit/499a0d5649b8d5d51caa6b04105e6e6ace4562f0

**Reply by mxcl** [reply]
> I will at least document all of the errors I get and their fixes, ok?

Yep that's super helpful, thank you
